File: JavaScript/6js 1/isValidWalk.js
/*
|--------------------------------------------------------------------------
| Take a Ten Minutes Walk
|--------------------------------------------------------------------------
|
*/

function isValidWalk(walk) {
   if (walk.length !== 10) return false;
   const x = { n: 0, s: 0, e: 0, w: 0 };
   for (let i of walk) {
      x[i]++;
   }
   return x["n"] === x["s"] && x["e"] === x["w"];
}


function isValidWalk(walk) {
   let dx = 0;
   let dy = 0;
   let dt = walk.length;
   for (let i = 0; i < walk.length; i++) {
      switch (walk[i]) {
         case "n":
            dy--;
            break;
         case "s":
            dy++;
            break;
         case "w":
            dx--;
            break;
         case "e":
            dx++;
            break;
      }
   }
   return dt === 10 && dx === 0 && dy === 0;
}


function isValidWalk(walk) {
   function count(val) {
      return walk.filter(function (a) {
         return a == val;
      }).length;
   }
   return walk.length == 10 && count("n") == count("s") && count("w") == count("e");
}


console.log(
   isValidWalk(["w", "e", "w", "e", "w", "e", "w", "e", "w", "e", "w", "e"]),
   "should return false"
);
console.log(isValidWalk(["w"]), "should return false");
console.log(isValidWalk(["n", "n", "n", "s", "n", "s", "n", "s", "n", "s"]), "should return false");

File: Python/6py 2/is_valid_walk.py
"""
|--------------------------------------------------------------------------
| Take a Ten Minutes Walk
|--------------------------------------------------------------------------
|
"""

def is_valid_walk(walk):
   if len(walk) != 10: return False
   x = {'n': 0, 's': 0, 'e': 0, 'w': 0}
   for i in walk: x[i] += 1
   return x['n'] == x['s'] and x['e'] == x['w']


def isValidWalk(walk):
   return len(walk) == 10 and walk.count('n') == walk.count('s') and walk.count('e') == walk.count('w')


def isValidWalk(walk):
   if len(walk) != 10: return False
   x, y = 0, 0
   for direction in walk:
      if direction == 'n': y += 1
      elif direction == 's': y -= 1
      elif direction == 'e': x += 1
      elif direction == 'w': x -= 1
      else: return False
   return x == 0 and y == 0


print(is_valid_walk(["w"]), "should return false")
print(is_valid_walk(["n", "n", "n", "s", "n", "s", "n", "s", "n", "s"]), "should return false")